File: src/features/recommendations/utils/match-score.test.ts
import { describe, it, expect } from "vitest";
import { deriveMatchScore } from "./match-score";

describe("deriveMatchScore", () => {
  it("clamps an AI-supplied score to 0-100", () => {
    expect(deriveMatchScore({ id: "x", match_score: 150 }).score).toBe(100);
    expect(deriveMatchScore({ id: "x", match_score: -10 }).score).toBe(0);
    expect(deriveMatchScore({ id: "x", match_score: 87.4 }).score).toBe(87);
  });

  it("falls back to a stable hash when no score is provided", () => {
    const a = deriveMatchScore({ id: "stable-id" });
    const b = deriveMatchScore({ id: "stable-id" });
    expect(a.score).toBe(b.score);
    expect(a.score).toBeGreaterThanOrEqual(78);
    expect(a.score).toBeLessThanOrEqual(97);
  });

  it("maps scores to tones at documented thresholds", () => {
    expect(deriveMatchScore({ id: "x", match_score: 92 }).tone).toBe("green");
    expect(deriveMatchScore({ id: "x", match_score: 90 }).tone).toBe("green");
    expect(deriveMatchScore({ id: "x", match_score: 80 }).tone).toBe("yellow");
    expect(deriveMatchScore({ id: "x", match_score: 75 }).tone).toBe("yellow");
    expect(deriveMatchScore({ id: "x", match_score: 50 }).tone).toBe("red");
  });

  it("ignores NaN/Infinity and falls back to hashed score", () => {
    const nanResult = deriveMatchScore({ id: "y", match_score: NaN });
    const infResult = deriveMatchScore({ id: "y", match_score: Infinity });
    expect(nanResult.score).toBeGreaterThanOrEqual(78);
    expect(infResult.score).toBeGreaterThanOrEqual(78);
  });
});

File: apps/mobile/test/match_score_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smart_advisor/features/recommendations/utils/match_score.dart';

void main() {
  group('deriveMatchScore', () {
    test('uses match_score, clamped and rounded, with tone thresholds', () {
      expect(deriveMatchScore(id: 'x', matchScore: 93).score, 93);
      expect(deriveMatchScore(id: 'x', matchScore: 93).tone, MatchTone.green);
      expect(deriveMatchScore(id: 'x', matchScore: 80).tone, MatchTone.yellow);
      expect(deriveMatchScore(id: 'x', matchScore: 50).tone, MatchTone.red);
      expect(deriveMatchScore(id: 'x', matchScore: 150).score, 100);
      expect(deriveMatchScore(id: 'x', matchScore: 88.6).score, 89);
    });

    test('falls back to a stable hash in 78..97 when score absent', () {
      final a = deriveMatchScore(id: 'rec-abc');
      final b = deriveMatchScore(id: 'rec-abc');
      expect(a.score, b.score, reason: 'deterministic per id');
      expect(a.score, inInclusiveRange(78, 97));
      // Many ids should spread across the 78..97 band (collisions on the
      // 20-value range are expected, so assert spread, not per-id distinctness).
      final spread = {
        for (var i = 0; i < 200; i++) deriveMatchScore(id: 'id-$i').score,
      };
      expect(spread.length, greaterThan(10));
      expect(spread.every((s) => s >= 78 && s <= 97), isTrue);
    });
  });
}
